### PR TITLE
Fix failing location activities

### LINKED ISSUE #301: Change old locations services in favor of new LocationEngine abstraction
Some activities which are using old locations services (`GeocodingWidgetActivity` and `GeocodingServiceActivity`) are crashing when calling `FusedLocationProviderService.requestLocationUpdates` or `FusedLocationProviderService.getLastLocation` methods.

> FATAL EXCEPTION: main
                                                                                     Process: com.mapbox.services.android.testapp, PID: 26935
                                                                                     java.lang.RuntimeException: Unable to start activity ComponentInfo{com.mapbox.services.android.testapp/com.mapbox.services.android.testapp.geocoding.GeocodingServiceActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'android.location.Location com.mapzen.android.lost.internal.FusedLocationProviderService.getLastLocation(com.mapzen.android.lost.api.LostApiClient)' on a null object reference
                                                                                         at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2416)
                                                                                         at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
                                                                                         at android.app.ActivityThread.-wrap11(ActivityThread.java)
                                                                                         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
                                                                                         at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                                         at android.os.Looper.loop(Looper.java:148)
                                                                                         at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                                         at java.lang.reflect.Method.invoke(Native Method)
                                                                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                                                      Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'android.location.Location com.mapzen.android.lost.internal.FusedLocationProviderService.getLastLocation(com.mapzen.android.lost.api.LostApiClient)' on a null object reference
                                                                                         at com.mapzen.android.lost.internal.FusedLocationProviderApiImpl.getLastLocation(FusedLocationProviderApiImpl.java:100)
                                                                                         at com.mapbox.services.android.testapp.geocoding.GeocodingServiceActivity.getLastLocation(GeocodingServiceActivity.java:155)
                                                                                         at com.mapbox.services.android.testapp.geocoding.GeocodingServiceActivity.onStart(GeocodingServiceActivity.java:148)
                                                                                         at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1237)
                                                                                         at android.app.Activity.performStart(Activity.java:6253)
                                                                                         at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2379)
                                                                                         at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476) 
                                                                                         at android.app.ActivityThread.-wrap11(ActivityThread.java) 
                                                                                         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344) 
                                                                                         at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                                         at android.os.Looper.loop(Looper.java:148) 
                                                                                         at android.app.ActivityThread.main(ActivityThread.java:5417) 
                                                                                         at java.lang.reflect.Method.invoke(Native Method) 
                                                                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
                                                                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 

Use new `LocationEngine` abstraction to fix those crashes.

### PR DESCRIPTION
**Pull request by Guardiola31337**
Fixes #301 

Added some `WeakReference`'s to avoid unwanted memory leaks.

👀  @zugaldia

### CODE REVIEW

File: mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingServiceActivity.java [hunk 1]
```diff
@@ -161,29 +169,42 @@ private void getLastLocation() {
...
-    if (lostApiClient.isConnected()) {
-      lostApiClient.disconnect();
+  public void onConnected() {
+    getLastLocation();
+  }
+
+  @Override
+  public void onLocationChanged(Location location) {
+    if (location != null) {
+      getLastLocation();
```
Comment (zugaldia) [inline mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingServiceActivity.java]: @Guardiola31337 We're ignoring the `location` value passed to `onLocationChanged` here.

Comment (Guardiola31337) [inline mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingServiceActivity.java]: Actually, we're not requesting location updates, so it's not necessary.
Problem. Are we going to leave `onLocationChanged` method empty? Should we segregate `LocationEngineListener` interface so client is not forced to depend on methods it does not use? What do you think? @zugaldia

Comment (zugaldia) [inline mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingServiceActivity.java]: @Guardiola31337 It's only one method so it doesn't bring a lot of overhead, it keeps things simple. I'd leave it like it right now. Also, it's more consistent with [`LocationListener`](https://developer.android.com/reference/android/location/LocationListener.html) this way.

**Comment by zugaldia** [comment]
@Guardiola31337 All good. If you want, and this is a minor thing, you could add a comment to the `onLocationChanged()` method simply saying that's unused on purpose for anyone reading the example.

**Pull request merged**